compiler/rustc_codegen_ssa/src/back/link.rs

```diff
@@ -1812,10 +1815,20 @@ fn add_local_native_libraries(
             None => continue,
         };
         match lib.kind {
-            NativeLibKind::Dylib | NativeLibKind::Unspecified => cmd.link_dylib(name),
-            NativeLibKind::Framework => cmd.link_framework(name),
-            NativeLibKind::StaticNoBundle => cmd.link_staticlib(name),
-            NativeLibKind::StaticBundle => cmd.link_whole_staticlib(name, &search_path),
+            NativeLibKind::Dylib { as_needed } => cmd.link_dylib(name, lib.verbatim, as_needed),
+            NativeLibKind::Unspecified => cmd.link_dylib(name, lib.verbatim, true),
+            NativeLibKind::Framework { as_needed } => cmd.link_framework(name, as_needed),
+            NativeLibKind::Static { bundle: true, .. } => {
+                cmd.link_whole_staticlib(name, lib.verbatim, &search_path);
+            }
+            NativeLibKind::Static { whole_archive: true, .. } => {
+                // Even though we pass `whole-archive` (or similar) to the linker here,
+                // gc-sections (or similar) may end up stripping away any unreferenced
+                // symbols. So let's try explicitly disabling that.
+                cmd.no_gc_sections();
```

Comment:
Are you sure that this is necessary and linker doesn't disable `--gc-sections` automatically when it sees `--whole-archive`? If it's necessary, then that's unfortunate.

Anyway, if it's necessary and `--gc-sections` is disabled before `link_whole_staticlib`, then it should be re-enabled after it.
Or not, because `--gc-sections` is not always enabled by default. (Pehaps `--push-state`/`--pop-state` can help here, although https://sourceware.org/binutils/docs/ld/Options.html doesn't list `--gc-sections` among the saved/restored flags.)
What is worse, `/OPT:NOREF,NOICF` on MSVC is global and applies to *all* libraries, not to this specific one. (But again, maybe it's not even necessary?)

Reply:
From reading LLD sources it looks like `--gc-sections` is a global flag affecting all files regarding of order in ld-like linkers as well.

Reply:
Yea, I did some quick testing and it seemed like `--whole-archive` by itself isn't enough unfortunately. And like you said, it's a global flag. :/